File: src/llm_usage/aggregation.py
from __future__ import annotations

from collections import defaultdict
from datetime import datetime
from zoneinfo import ZoneInfo

from .identity import build_row_key
from .models import AggregateRecord, UsageEvent


def aggregate_events(
    events: list[UsageEvent],
    user_hash: str,
    timezone_name: str,
    now: datetime | None = None,
) -> list[AggregateRecord]:
    tz = ZoneInfo(timezone_name)
    curr = (now or datetime.now(tz)).astimezone(tz)

    buckets: dict[tuple[str, str, str, str], dict[str, object]] = defaultdict(
        lambda: {
            "input": 0,
            "cache": 0,
            "output": 0,
            "model": "unknown",
            "model_time": None,
            "session_fingerprint": None,
            "source_host_hash": "",
        }
    )

    for event in events:
        local_date = event.event_time.astimezone(tz).date().isoformat()
        identity = event.session_fingerprint.strip() if event.session_fingerprint else f"model:{event.model}"
        key = (local_date, event.tool, event.source_host_hash, identity)
        buckets[key]["input"] += max(0, event.input_tokens)
        buckets[key]["cache"] += max(0, event.cache_tokens)
        buckets[key]["output"] += max(0, event.output_tokens)
        buckets[key]["source_host_hash"] = event.source_host_hash
        if event.session_fingerprint:
            buckets[key]["session_fingerprint"] = event.session_fingerprint.strip()
        if event.model != "unknown":
            model_time = buckets[key]["model_time"]
            if not isinstance(model_time, datetime) or event.event_time >= model_time:
                buckets[key]["model"] = event.model
                buckets[key]["model_time"] = event.event_time

    out: list[AggregateRecord] = []
    updated_at = curr.isoformat()
    for (date_local, tool, source_host_hash, _identity), sums in sorted(buckets.items()):
        model = str(sums["model"])
        session_fingerprint = sums["session_fingerprint"]
        if not isinstance(session_fingerprint, str):
            session_fingerprint = None
        out.append(
            AggregateRecord(
                date_local=date_local,
                user_hash=user_hash,
                source_host_hash=source_host_hash,
                tool=tool,
                model=model,
                input_tokens_sum=sums["input"],
                cache_tokens_sum=sums["cache"],
                output_tokens_sum=sums["output"],
                row_key=build_row_key(
                    user_hash,
                    source_host_hash,
                    date_local,
                    tool,
                    model,
                    session_fingerprint=session_fingerprint,
                ),
                updated_at=updated_at,
            )
        )
    return out

File: node/src/core/aggregation.js
import { buildRowKey, resolveIdentity } from "./identity.js";
import { toDate } from "./models.js";
import { formatDateLocal, formatIsoInTimeZone } from "./time.js";

function clampTokens(value) {
  return Math.max(0, Number(value || 0));
}

export function aggregateEvents(events, { userHash, timeZone, now = new Date() }) {
  const buckets = new Map();

  for (const rawEvent of events) {
    const eventTime = toDate(rawEvent.eventTime);
    const sourceHostHash = rawEvent.sourceHostHash || "";
    const localDate = formatDateLocal(eventTime, timeZone);
    const identity = resolveIdentity(rawEvent.model, rawEvent.sessionFingerprint);
    const bucketKey = `${localDate}\u0000${rawEvent.tool}\u0000${sourceHostHash}\u0000${identity}`;
    const existing =
      buckets.get(bucketKey) ||
      {
        date_local: localDate,
        tool: rawEvent.tool,
        source_host_hash: sourceHostHash,
        input_tokens_sum: 0,
        cache_tokens_sum: 0,
        output_tokens_sum: 0,
        model: "unknown",
        model_time: null,
        session_fingerprint: null,
      };

    existing.input_tokens_sum += clampTokens(rawEvent.inputTokens);
    existing.cache_tokens_sum += clampTokens(rawEvent.cacheTokens);
    existing.output_tokens_sum += clampTokens(rawEvent.outputTokens);
    existing.source_host_hash = sourceHostHash;

    if (typeof rawEvent.sessionFingerprint === "string" && rawEvent.sessionFingerprint.trim()) {
      existing.session_fingerprint = rawEvent.sessionFingerprint.trim();
    }

    if (rawEvent.model !== "unknown" && (!existing.model_time || eventTime >= existing.model_time)) {
      existing.model = rawEvent.model;
      existing.model_time = eventTime;
    }

    buckets.set(bucketKey, existing);
  }

  const updatedAt = formatIsoInTimeZone(toDate(now), timeZone);
  return [...buckets.values()]
    .sort((left, right) => {
      return (
        left.date_local.localeCompare(right.date_local) ||
        left.tool.localeCompare(right.tool) ||
        left.source_host_hash.localeCompare(right.source_host_hash) ||
        resolveIdentity(left.model, left.session_fingerprint).localeCompare(
          resolveIdentity(right.model, right.session_fingerprint),
        )
      );
    })
    .map((bucket) => ({
      date_local: bucket.date_local,
      user_hash: userHash,
      source_host_hash: bucket.source_host_hash,
      tool: bucket.tool,
      model: bucket.model,
      input_tokens_sum: bucket.input_tokens_sum,
      cache_tokens_sum: bucket.cache_tokens_sum,
      output_tokens_sum: bucket.output_tokens_sum,
      row_key: buildRowKey({
        userHash,
        sourceHostHash: bucket.source_host_hash,
        dateLocal: bucket.date_local,
        tool: bucket.tool,
        model: bucket.model,
        sessionFingerprint: bucket.session_fingerprint,
      }),
      updated_at: updatedAt,
    }));
}
